ex3/m_depontes/main.py

```diff
@@ -0,0 +1,473 @@
+        transform: 学習した変換行列によって特徴量ベクトルを射影する関数
+        explained_variance_ratio: 累積寄与率を計算する関数
+    属性：
+        n_components(int): 主成分の数
+        A(np.ndarray): 変換行列, shape = (次元数, 次元数)
+        eigenvalues(np.ndarray): 固有値, shape = (次元数,)
+        eigenvectors(np.ndarray): 固有ベクトル, shape = (次元数, 次元数)
+        means(np.ndarray): 期待値ベクトル, shape = (次元数)
+    """
+
+    def __init__(
+        self,
+        n_components: int = 2,
+    ):
+        """初期化関数.
+
+        入力：
+            n_components(int): 主成分の数
+        出力：
+            None
+        """
+        self.n_components = n_components
+        self.A = None
+        self.eigenvalues = None
+        self.eigenvectors = None
+        self.means = None
+
+    def compute_S(
+        self,
+        X: np.ndarray,
+        means: np.ndarray,
+    ) -> np.ndarray:
+        """観測値ベクトルの集合およびその期待値ベクトルから変動行列を計算する関数.
+
+        入力：
+            X(np.ndarray): 観測値ベクトルの集合, shape = (観測値ベクトルの数, 次元数)
+            means(np.ndarray): 期待値ベクトル, shape = (次元数)
+
+        出力：
+            S(np.ndarray): 変動行列, shape = (次元数, 次元数)
+        """
+        S = 0
+        for i in range(len(X)):
+            S += np.outer((X[i] - means), (X[i] - means).T)
+        return S
+
+    def compute_sorted_eigen(
+        self,
+        cov_matrix: np.ndarray,
+    ) -> tuple[np.ndarray, np.ndarray]:
+        """分散共分散行列から固有値および固有ベクトルを計算し，それぞれ固有値の降順に並べて返す関数.
+
+        入力：
+            cov_matrix(np.ndarray): 分散共分散行列, shape = (次元数, 次元数)
+        出力：
+            eigenvalues(np.ndarray): 固有値, shape = (次元数,)
+            eigenvectors(np.ndarray): 固有ベクトル, shape = (次元数, 次元数)
+        """
+        eigenvalues, eigenvectors = np.linalg.eig(cov_matrix)
+        index = np.argsort(eigenvalues)[::-1]
+        self.eigenvalues = eigenvalues[index]
+        self.eigenvectors = eigenvectors[:, index]
+
+        return self.eigenvalues, self.eigenvectors
+
+    def fit(
+        self,
+        X: np.ndarray,
+    ) -> np.ndarray:
+        """分散最大基準に基づき，変換行列を学習する関数.
+
+        入力：
+            X(np.ndarray): 観測値ベクトルの集合, shape = (観測値ベクトルの数, 次元数)
+        出力：
+            A(np.ndarray): 変換行列, shape = (次元数, 次元数)
+        """
+        # 特徴ベクトルの総数を取得
+        n = len(X)
+
+        # 特徴ベクトルの分散共分散行列を計算
+        self.means = X.mean(axis=1)
+        centered_data = X - self.means[:, np.newaxis]
+        S = self.compute_S(centered_data.T, np.zeros(centered_data.shape[0])) / n
+
+        # 分散共分散行列から固有値及び固有値ベクトルを計算し，固有値の大きい順に整列
+        eigenvalues, eigenvectors = self.compute_sorted_eigen(S)
+
+        # 上位n_components個の固有値ベクトルを変換行列として抽出
+        self.A = eigenvectors[: self.n_components]
+        return self.A
+
+    def transform(
+        self,
+        X: np.ndarray,
+    ) -> np.ndarray:
+        """学習した変換行列によって特徴量ベクトルを射影する関数.
+
+        入力：
+            X(np.ndarray): 観測値ベクトルの集合, shape = (観測値ベクトルの数, 次元数)
+        出力：
+            _(np.ndarray): 射影された特徴量ベクトル, shape = (観測値ベクトルの数, 次元数)
+        """
+        return np.dot(self.A, X)
```

Comment:
```suggestion
        出力：
            _(np.ndarray): 射影された特徴量ベクトル, shape = (観測値ベクトルの数, 主成分数)
        """
        return np.dot(self.A, X).T
```
`np.dot(self.A, X)`のshapeは(主成分数, 観測値ベクトルの数)でdocstringと逆です。